dash/testing/browser.py

```diff
@@ -0,0 +1,197 @@
+import logging
+import percy
+
+from selenium import webdriver
+from selenium.webdriver.support import expected_conditions as EC
+from selenium.webdriver.common.by import By
+from selenium.webdriver.support.wait import WebDriverWait
+from selenium.webdriver.chrome.options import Options
+from selenium.webdriver.common.desired_capabilities import DesiredCapabilities
+from selenium.common.exceptions import WebDriverException, TimeoutException
+
+from dash.testing.wait import text_to_equal
+from dash.exceptions import DashAppLoadingError
+
+logger = logging.getLogger(__name__)
+
+
+class Browser:
+    def __init__(self, browser, remote=None, wait_timeout=10):
+        self._browser = browser.lower()
+        self._wait_timeout = wait_timeout
+
+        self._driver = self.get_webdriver(remote)
+        self._driver.implicitly_wait(2)
+
+        self._wd_wait = WebDriverWait(
+            driver=self.driver, timeout=wait_timeout, poll_frequency=0.2
+        )
+        self._last_ts = 0
+        self._url = None
+
+        self.percy_runner = percy.Runner(
+            loader=percy.ResourceLoader(
+                webdriver=self.driver,
+                base_url="/assets",
+                root_dir="tests/assets",
+            )
+        )
+        self.percy_runner.initialize_build()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, traceback):
+        try:
+            self.driver.quit()
+            self.percy_runner.finalize_build()
+        except WebDriverException:
+            logger.exception("webdriver quit was not successfully")
+        except percy.errors.Error:
+            logger.exception("percy runner failed to finalize properly")
+
+    def percy_snapshot(self, name=""):
+        snapshot_name = "{} - py{}.{}".format(
+            name, sys.version_info.major, sys.version_info.minor
+        )
+        logger.info("taking snapshot name => %s", snapshot_name)
+        self.percy_runner.snapshot(name=snapshot_name)
+
+    def _wait_for(self, method, args, timeout, msg):
+        """abstract generic pattern for explicit webdriver wait"""
+        _wait = (
+            self._wd_wait
+            if timeout is None
+            else WebDriverWait(self.driver, timeout)
+        )
+        return _wait.until(method(*args), msg)
+
+    # keep these two wait_for API for easy migration
+    def wait_for_element_by_css_selector(self, selector, timeout=None):
+        return self._wait_for(
+            EC.presence_of_element_located,
+            ((By.CSS_SELECTOR, selector),),
+            timeout,
+            "cannot find_element using the css selector",
+        )
+
+    def wait_for_text_to_equal(self, selector, text, timeout=None):
+        return self._wait_for(
+            text_to_equal,
+            (selector, text),
+            timeout,
+            "cannot wait until element contains expected text {}".format(text),
+        )
+
+    def wait_until_server_is_ready(self, timeout=10):
```

Comment:
`wait_until_page_is_ready`? or even just `wait_for_page`?

Reply:
yep, I don't like the current api name, `wait_for_page` is cool for me